Circleci 2.1 config win cache

filipesilva opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
CircleCI config uses anchors a bit too extensively. 

CircleCI windows jobs don't use cache or workspaces.


## What is the new behavior?
CircleCI config uses Executors and Commands present in 2.1 config.

CircleCI windows jobs use cache or workspaces.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

FW repository version of https://github.com/angular/angular-cli/commit/38f5eeeebfca6af13f28ed7ea930279ebf8e1261#diff-1d37e48f9ceff6d8030570cd36286a61 and https://github.com/angular/angular-cli/pull/15038.

Requires CircleCI pipelines to be enable in order to support 2.1 configuration version.

filipesilva committed:
ci: use CircleCI executors

Executors let you define custom execution contexts for jobs.

See https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-executors for more information.

filipesilva committed:
ci: keep cache key fallback in a var

This way it's right next to the original key and it's easier to keep them in sync.

filipesilva committed:
ci: use CircleCI commands

When we needed to run multiple commands in a reusable fashion, we needed to make a giant run block with multiple things inside. Using custom commands gives us a better way to do this.

See https://circleci.com/docs/2.0/reusing-config/#authoring-reusable-commands for more info.

filipesilva committed:
ci: use workspaces and cache for Windows jobs

Making use of these functionalities lets our Windows jobs be more similar to the Linux ones, and draw common artifacts from them too.

filipesilva committed:
ci: use CircleCI parameterized jobs

Parameterized jobs lets us reduce duplication of very similar jobs.

See https://circleci.com/docs/2.0/reusing-config/#authoring-parameterized-jobs for more info.

mary-poppins commented:
You can preview fc384a1 at https://pr32745-fc384a1.ngbuilds.io/.

mary-poppins commented:
You can preview eedff87 at https://pr32745-eedff87.ngbuilds.io/.

filipesilva commented:
Note to reviewers: the windows jobs usually don't run on PRs, but I forced them to run in https://circleci.com/workflow-run/01d13c10-b1d3-4f94-8cf5-82ae90958c84 to show it still works correctly.

mary-poppins commented:
You can preview b441836 at https://pr32745-b441836.ngbuilds.io/.

gkalpak commented on .circleci/config.yml:
```diff
@@ -430,45 +419,33 @@ jobs:
+            - run: mv dist/packages-dist-ivy-aot dist/packages-dist
+        # Run examples tests. The "CIRCLE_NODE_INDEX" will be set if "parallelism" is enabled.
+        # Since the parallelism is set to "5", there will be five parallel CircleCI containers.
+        # with either "0", "1", etc" as node index. This can be passed to the "--shard" argument.
```
`with either "0", "1", etc"` --> `with "0", "1", etc` (?)

filipesilva replied:
Fixed

gkalpak commented on .circleci/config.yml:
```diff
@@ -430,45 +419,33 @@ jobs:
-        # with either "0", "1" or "2" as node index. This can be passed to the "--shard" argument.
-      - run: yarn --cwd aio example-e2e --setup --local --ivy --cliSpecsConcurrency=5 --shard=${CIRCLE_NODE_INDEX}/${CIRCLE_NODE_TOTAL} --retry 2
+      - when:
+          condition: <<parameters.ivy>>
```
Super-nit: `<<parameters.ivy>>` --> `<< parameters.ivy >>` (for consistency 😇)

filipesilva replied:
Fixed

gkalpak commented on aio/package.json:
```diff
@@ -19,10 +19,10 @@
+    "prebuild-local-with-ivy": "yarn setup-local && node scripts/switch-to-ivy",
+    "build-local-with-ivy": "yarn ~~build",
+    "prebuild-local-with-ivy-ci": "yarn setup-local --no-build-packages && node scripts/switch-to-ivy",
+    "build-local-with-ivy-ci": "yarn ~~build --progress=false",
```
The script names also need to be updated in [aio/README.md][1].

[1]: https://github.com/angular/angular/blob/edfb0e8c8/aio/README.md

filipesilva replied:
Thank you, I didn't checked there. Fixed now

gkalpak commented on .circleci/config.yml:
```diff
@@ -859,7 +835,8 @@ workflows:
             - setup
-      # Windows jobs very slow so we run it on non-PRs only for now.
+      # Windows Jobs
+      # These are very slow so we run it on non-PRs only for now.
```
Typo: `it` --> `them`

filipesilva replied:
Fixed

gkalpak reviewed (changes requested):
This is awesome 😍
Great job :tada: I love it! ❤️ 

I left a few minor comments. I guess it remains to be seen whether the caching works on Windows :grin:

The links to docs in comments are just great 👌 💯
One last nit: I think the commit messages could have a little more info in the body (expanding a bit on the commit header and maybe pointing to relevant docs).

gkalpak commented:
BTW, I just ran into this neat trick to avoid the annoying `var_X:` keys when defining anchors: https://github.com/integratedexperts/drupal-dev/blob/11ed22b48/.circleci/config.yml#L13-L23

filipesilva commented on .circleci/config.yml:
```diff
@@ -26,58 +26,35 @@ var_4: &cache_key_fallback v3-angular-node-10.16-
+      ignore:
+        - /pull\/.*/
+
+# TODO: move these anchors into a windows setup job when CircleCI fixes cache/workflows on win.
```
They were moved in subsequent commits within this PR. I left the comment in case some commits were cherry-picked to explain why the windows anchors weren't refactored along the other anchors that were refactored in that commit.

filipesilva commented on .circleci/config.yml:
```diff
@@ -238,16 +245,16 @@ jobs:
       # **NOTE 1 **: Folders persisted here should be kept in sync with `var_13: &attach_workspace`.
```
Now that you drew my attention to it, I think this comment doesn't make sense. You can't selectively attach parts of a workspace. The only thing that matters is its root.

filipesilva commented on .circleci/config.yml:
```diff
@@ -7,18 +7,11 @@
-# Note that the browser docker image comes with Chrome and Firefox preinstalled. This is just
-# needed for jobs that run tests without Bazel. Bazel runs tests with browsers that will be
-# fetched by the Webtesting rules. Therefore for jobs that run tests with Bazel, we don't need a
-# docker image with browsers pre-installed.
```
It was separated from bottom bit and moved under the `browsers-executor`, because that's where it applies.

filipesilva commented on .circleci/config.yml:
```diff
@@ -195,10 +163,62 @@ var_14: &notify_dev_infra_on_fail
+#             `CI_CHROMEDRIVER_VERSION_ARG` env var (in `.circleci/env.sh`) points to a ChromeDriver
+#             version that is compatible with the Chrome version in the image.
+executors:
+  action-executor:
```
I didn't think too much about that name actually, just copied it over from the CLI PR.

What do you think would be a better name? Maybe `default-executor`.

gkalpak replied:
`default-executor` sgtm.

filipesilva replied:
done

filipesilva commented on .circleci/config.yml:
```diff
@@ -18,14 +18,16 @@ version: 2.1
 # 1) yarn lock file changes --> cached "node_modules" are different.
 # 2) bazel repository definitions change --> cached bazel repositories are different.
-# **NOTE 1 **: If you change the cache key prefix, also sync the restore_cache fallback to match.
+# **NOTE 1 **: If you change the cache key prefix, also sync the fallback_cache_key to match.
```
It should be `cache_key_fallback`, I forgot to change it. Fixed.

filipesilva commented on .circleci/config.yml:
```diff
@@ -26,58 +26,35 @@ var_4: &cache_key_fallback v3-angular-node-10.16-
-      echo "export GOOGLE_APPLICATION_CREDENTIALS=/home/circleci/.gcp_credentials" >> $BASH_ENV
-      ./.circleci/setup-rbe.sh .bazelrc.user
+# Workspace initially persisted by the `setup` job, and then enhanced by `test_aio` and
+# `build-npm-packages`.
```
Fixed

filipesilva commented on .circleci/config.yml:
```diff
@@ -26,58 +26,35 @@ var_4: &cache_key_fallback v3-angular-node-10.16-
+# `build-npm-packages`.
+# https://circleci.com/docs/2.0/workflows/#using-workspaces-to-share-data-among-jobs
+# https://circleci.com/blog/deep-diving-into-circleci-workspaces/
+var_7: &workspace_location ~/
```
Which are arguably the most important places for it to be :D
Fixed

mary-poppins commented:
You can preview 62e29ed at https://pr32745-62e29ed.ngbuilds.io/.

filipesilva commented:
@gkalpak thanks for making the comments in the individual commits, that made it much easier to make fixups. I have added more context on the commit messages now.

Regarding the trick for avoiding `var_X`: if I remember correctly, we actually introduced the `anchor_x` and `var_x` names in https://github.com/angular/angular/pull/18101/files so that it was clear that these things were anchors and vars and whatnot, because yaml is kinda terse and confusing on this topic. So I'm not sure we actually want to drop it.

filipesilva committed:
ci: refactor notification commands into a single parametrized one

notify_caretaker_on_fail and notify_dev_infra_on_fail are the same, except for the url they notify.

filipesilva commented on .circleci/config.yml:
```diff
@@ -216,16 +125,114 @@ executors:
+          # The URL comes from https://angular-team.slack.com/apps/A0F7VRE7N-circleci.
+          command: |
+            notificationJson="{\"text\":\":x: \`$CIRCLE_JOB\` job for $CIRCLE_BRANCH branch failed on build $CIRCLE_BUILD_NUM: $CIRCLE_BUILD_URL :scream:\"}"
+            curl --request POST --header "Content-Type: application/json" --data "$notificationJson" $SLACK_DEV_INFRA_CI_FAILURES_WEBHOOK_URL
```
Done as well.

mary-poppins commented:
You can preview fe94a4e at https://pr32745-fe94a4e.ngbuilds.io/.

gkalpak commented:
> Regarding the trick for avoiding `var_X`: if I remember correctly, we actually introduced the `anchor_x` and `var_x` names in https://github.com/angular/angular/pull/18101/files so that it was clear that these things were anchors and vars and whatnot

Putting them in an object keyed as `anchors` or `vars` is supposed to give a hint (but not as clear as having it in front of every anchor. 

I find YAML so confusing/unreadable anyway; I don't mind either way 😛

gkalpak commented on .circleci/config.yml:
```diff
@@ -178,27 +178,19 @@ commands:
+      # `webhook_url_env_var` are secret env vars defined in CircleCI project settings.
+      # The URLs come from https://angular-team.slack.com/apps/A0F7VRE7N-circleci.
+      webhook_url_env_var:
+        type: string
```
The docs are not clear on why, but [env_var_name][1] seems to be a better type for this parameter.

I would also feel better, if we only passed the variable name (without the `$`) to make it clear that the value (which should remain secret) is not exposed in the config.

[1]: https://circleci.com/docs/2.0/reusing-config/#environment-variable-name

filipesilva replied:
Done

gkalpak reviewed (changes requested):
Thx for adding the info to commit messages :+1:
Left one last comment (and there is also https://github.com/angular/angular/pull/32745#discussion_r328207559).

Otherwise lgtm ✨

mary-poppins commented:
You can preview 6507bac at https://pr32745-6507bac.ngbuilds.io/.

mary-poppins commented:
You can preview ddbe3d6 at https://pr32745-ddbe3d6.ngbuilds.io/.

josephperrott commented on .circleci/config.yml:
```diff
@@ -293,16 +312,17 @@ jobs:
-    resource_class: xlarge
+    executor:
+      name: action-executor
+      # In order to avoid the bottleneck of having a slow host machine, we acquire a better
```
I'm a little confused by this comment.  From the comment it seems like we should not be requesting an xlarge resource class machine?

filipesilva replied:
I don't quite follow. The comment seems to be saying that an xlarge machine is needed because otherwise it would be a bottleneck.

josephperrott reviewed (approved):
Overall LGTM

mary-poppins commented:
You can preview a632487 at https://pr32745-a632487.ngbuilds.io/.

atscott commented:
Hi @filipesilva - There was a merge conflict with the patch branch. Can you create a new PR that targets path only and resolve those conflicts?

filipesilva commented:
@atscott patch version is https://github.com/angular/angular/pull/32982
